README.md
A simple dart package to convert large numbers to a human readable format. 1278 to 1.2K instead, for example.

## Features

Represents large numbers in terms of K, M, B, T etc. (1234 -> 1.2K)

### humanizeInt

| Integer   | Humanized |
| --------- | --------- |
| 1278      | 1.2K      |
| 250       | 250       |
| 12340000  | 12M       |
| 123400000 | 120M      |

### humanizeIntInd

Represents large numbers in terms of K, L, Cr (Indian Numbering System) etc. (1234 -> 1.2K)

| Integer   | Humanized |
| --------- | --------- |
| 1278      | 1.2K      |
| 250       | 250       |
| 123400    | 1.2L      |
| 123400000 | 12Cr      |

## Usage

### humanizeInt

```dart
  final n1 = humanizeInt(1234); // 1.2K
  final n2 = humanizeInt(12340000); // 12M
```

### humanizeIntInd

```dart
  final n1 = humanizeIntInd(1234); // 1.2K
  final n2 = humanizeIntInd(12340000); // 1.2Cr
```

## Additional information

This package only offers this niche functionality. For more general use cases, checkout

- [humanize](https://pub.dev/packages/humanize)
- [humanizer](https://pub.dev/packages/humanizer)


CHANGELOG.md
## 1.0.0

- Initial version.
- humanizeInt Function added

## 1.0.1

- Added support for Indian Numbering System


test/humanize_big_int_test.dart
import 'package:humanize_big_int/humanize_big_int.dart';
import 'package:test/test.dart';

void main() {
  group('Tests for humanizeInt', () {
    test("Returns same if small number", () => expect("234", humanizeInt(234)));
    test("1,000s without decimal", () => expect("1K", humanizeInt(1000)));
    test("1,000s with decimal", () => expect("1.2K", humanizeInt(1234)));
    test("10,000s without decimal", () => expect("12K", humanizeInt(12340)));
    test("100,000s without decimal", () => expect("123K", humanizeInt(123400)));
    test(
        "1,000,000s without decimal", () => expect("1M", humanizeInt(1000000)));
    test("1,000,000s with decimal", () => expect("1.2M", humanizeInt(1234000)));
    test("10,000,000s without decimal",
        () => expect("12M", humanizeInt(12340000)));
    test("100,000,000s without decimal",
        () => expect("123M", humanizeInt(123400000)));
    test("1,000,000,000s with decimal",
        () => expect("1.2B", humanizeInt(1234000000)));
    test("1,000,000,000s without decimal",
        () => expect("1B", humanizeInt(1000000000)));
    test("10,000,000,000s without decimal",
        () => expect("12B", humanizeInt(12340000000)));
    test("100,000,000,000s without decimal",
        () => expect("123B", humanizeInt(123400000000)));
    test("1,000,000,000,000s with decimal",
        () => expect("1.2T", humanizeInt(1234000000000)));
    test("1,000,000,000,000s without decimal",
        () => expect("1T", humanizeInt(1000000000000)));
    test("10,000,000,000,000s without decimal",
        () => expect("12T", humanizeInt(12340000000000)));
    test("100,000,000,000,000s without decimal",
        () => expect("123T", humanizeInt(123400000000000)));
    test("100,000,000,000,000,000s ",
        () => expect("1Q", humanizeInt(1234000000000000)));
  });

  group("Test for huamizeIntInd", () {
    test("Returns same if small number",
        () => expect("234", humanizeIntInd(234)));
    test("1,000s without decimal", () => expect("1K", humanizeIntInd(1000)));
    test("1,000s with decimal", () => expect("1.2K", humanizeIntInd(1234)));
    test("10,000s without decimal", () => expect("12K", humanizeIntInd(12340)));
    test(
        "1,00,000s with decimal", () => expect("1.2L", humanizeIntInd(123400)));
    test("1,00,000s without decimal",
        () => expect("1L", humanizeIntInd(100000)));
    test("10,00,000s without decimal",
        () => expect("12L", humanizeIntInd(1234000)));
    test("1,00,00,000s with decimal",
        () => expect("1.2Cr", humanizeIntInd(12340000)));
    test("1,00,00,000s without decimal",
        () => expect("12Cr", humanizeIntInd(123400000)));
    test("100,00,00,000s with decimal",
        () => expect("123Cr", humanizeIntInd(1234000000)));
  });
}


lib/src/humanize_big_int_base.dart
import 'dart:math';

/// Function that converts large numbers to humanized form \
/// Returns same if number < 1000 \
/// Else Returns rounded number with appropriate suffix \
/// Eg: \
/// \
/// 234        ->   234 \
/// 1000       ->   1K \
/// 1234       ->   1.2K  \
/// 12340      ->   12K   \
/// 123400     ->   123K    \
/// 1234000    ->   1.2M    \
/// ....
String humanizeInt(int n) {
  // < 1000
  if (n < 1000) return n.toString();

  // 1,000s
  if (n < 10000) {
    final s = (n / 1000).toStringAsFixed(1);
    if (s[s.length - 1] == '0') {
      return s.substring(0, s.length - 2) + "K";
    }
    return s + "K";
  }
  if (n < 1000000) {
    return (n / 1000).toStringAsFixed(0) + "K";
  }

  // 1,000,000s
  if (n < 10000000) {
    final s = (n / pow(1000, 2)).toStringAsFixed(1);
    if (s[s.length - 1] == '0') {
      return s.substring(0, s.length - 2) + "M";
    }
    return s + "M";
  }
  if (n < 1000000000) {
    return (n / pow(1000, 2)).toStringAsFixed(0) + "M";
  }

  // 1,000,000,000s
  if (n < 10000000000) {
    final s = (n / pow(1000, 3)).toStringAsFixed(1);
    if (s[s.length - 1] == '0') {
      return s.substring(0, s.length - 2) + "B";
    }
    return s + "B";
  }
  if (n < 1000000000000) {
    return (n / pow(1000, 3)).toStringAsFixed(0) + "B";
  }

  // 1,000,000,000,000s
  if (n < 10000000000000) {
    final s = (n / pow(1000, 4)).toStringAsFixed(1);
    if (s[s.length - 1] == '0') {
      return s.substring(0, s.length - 2) + "T";
    }
    return s + "T";
  }
  if (n < 1000000000000000) {
    return (n / pow(1000, 4)).toStringAsFixed(0) + "T";
  }

  // >= 1,000,000,000,000,000s

  return (n / pow(1000, 5)).toStringAsFixed(0) + "Q";
}

/// Function that converts large numbers to humanized form \
/// Returns same if number < 1000 \
/// Else Returns rounded number with appropriate suffix in Indian numbering system (Thousand, Lakhs and Crores)\
/// Eg: \
/// \
/// 234        ->   234 \
/// 1000       ->   1K \
/// 1234       ->   1.2K  \
/// 12340      ->   12K   \
/// 1234000    ->   12L    \
/// 12340000   ->   1.2Cr
/// ....
String humanizeIntInd(int n) {
  // < 1000
  if (n < 1000) return n.toString();

  // 1,000s
  if (n < 10000) {
    final s = (n / 1000).toStringAsFixed(1);
    if (s[s.length - 1] == '0') {
      return s.substring(0, s.length - 2) + "K";
    }
    return s + "K";
  }
  if (n < 100000) {
    return (n / 1000).toStringAsFixed(0) + "K";
  }

  // 1,00,000s
  if (n < 1000000) {
    final s = (n / 100000).toStringAsFixed(1);
    if (s[s.length - 1] == '0') {
      return s.substring(0, s.length - 2) + "L";
    }
    return s + "L";
  }
  if (n < 10000000) {
    return (n / 100000).toStringAsFixed(0) + "L";
  }

  // 1,00,00,000s
  if (n < 100000000) {
    final s = (n / 10000000).toStringAsFixed(1);
    if (s[s.length - 1] == '0') {
      return s.substring(0, s.length - 2) + "Cr";
    }
    return s + "Cr";
  }
  if (n < 1000000000) {
    return (n / 10000000).toStringAsFixed(0) + "Cr";
  }

  // >= 100,00,00,000s

  return (n / 10000000).toStringAsFixed(0) + "Cr";
}
